### DISCUSSION TITLE
Send data between LEGO hubs and Raspberry Pi Pico

### DISCUSSION THREAD

**Discussion by KWSmit** [pybricks/support#1599] [Show and tell]
Pybricks uses Bluetooth BLE data broadcasting for sending information between LEGO hubs. The same technique can be used to send data between LEGO hubs and Raspberry Pi Pico. The good news is that it's not that difficult, thanks to Pybricks. I made a tutorial on [Github](https://github.com/KWSmit/ble-pico-pybricks). This way you can use a lot more sensors in your Pybricks projects. I'm thinking of using rfid-tags to distinguish which train enters the station :smile: Any other ideas?

**Comment by jaguilar** [comment]
Something along the lines of your plan to use the RFID tag to determine the position of a train is on my docket of things to do as well. It might be interesting to set the Pico up to command the Lego Hub via UART (or vice versa) so you only have to write the control program on one of the two devices, and the other one can be a simple listener program that never needs to change. Personally, I would run the control program on the Pico, since I'm getting into the C/C++-on-MCUs game.

**Reply by KWSmit** [reply]
That is something I have in my mind as well. Run a basic program on the train-hubs and the hubs that control switches and command them via UART. But I was thinking of controlling them from my PC. This gives you a lot of flexibility to change the track-layout or add trains.

**Reply by laurensvalk** [reply]
**@novakasa has entered the chat** :smile: See also
- https://www.youtube.com/watch?v=UZj41oFs1KA
- https://github.com/Novakasa/brickrail

**Reply by jaguilar** [reply]
Hah, amazing. That's a beautiful looking program too. I doubt I would have the patience to make such a nice UI!

**Comment by laurensvalk** [comment]
This looks great, thanks for sharing @KWSmit !

**Comment by JorgePe** [comment]
Thanks for sharing this.
I was searching for a quick and cheap solution to integrate a LEGO hub with my new OpenHAB installation and now I have a Raspberry Pi Pico W as a mqtt2pybricks gateway. For now just a proof of concept - created a "Generic MQTT Thing" on the OpenHAB dashboard with one switch channel, defined a topic and the payloads to the ON and OFF commands and now I can turn a LEGO motor ON or OFF and integrate this "thing" with my other house "things" (so I can make a LEGO robot act according to the humidity level of a room or the status of a window...)

**Reply by KWSmit** [reply]
@JorgePe Sounds cool :smile: . Will there be a video of this?

**Reply by JorgePe** [reply]
Will be. When I get something better than just a motor. And also when I get my Pico gateway stable, it crashes after just 2 hours of working, not good for a domotic system.

**Reply by JorgePe** [reply]
OK, not production-grade but I think it serves as a demo: a fingerbot to control my bathroom lights ON/OFF.

The bathroom fan has no independent control so I have to turn the lights ON to circulate the air and control humidity. We moved to a reconditioned house one year ago and we're facing some humidity problems (it seems to be a general problem in this neighborhood). So I got some dehumidifiers and some humidity sensors and integrated everything with OpenHAB so I can turn them all ON with a time schedule or with a humidity threshold.

Next I will insert  this fingerbot on the same rules of the dehumidifiers.

The Pico is in the room at front. I had it in another room (~10 meter) and sometimes the hub didn't receive the broadcast command.

I'm using the hub with batteries. If this gets approved by The Wife will make a power supply but for now the idea is to mount the fingerbot on the wall just when needed.

A link to the video:  https://youtu.be/EwQePDEx5PI

And 2 lousy pictures:
![The fingerbot](https://64.media.tumblr.com/13d6355b5fb7437278a291820d15fbe8/4f857a15594a97fa-d2/s2048x3072/173c18783dc45eba6a7348d700d0e4da8c041db8.jpg)
![The mqtt2pybricks gateway](https://64.media.tumblr.com/7ff14c99d14eb538974ad26d7a3d3581/4f857a15594a97fa-33/s2048x3072/5df86f501b4304183086babc48f02d5d8077dc49.jpg)

**Reply by KWSmit** [reply]
Cool project solving real life issues 😀

**Reply by JorgePe** [reply]
Thanks.

I already had a "switch-bot bot":
![A switch-bot bot](https://www.switch-bot.com/cdn/shop/products/switchbot-bot-781957.jpg?v=1701944767)

it's great to control appliances that have a digital power button (one of my dehumidifiers) where you cannot leave it just permanently "ON" and control from a smart power plug. But not so great for two-state switches like this one. It is also BLE but i found some python examples and made my own gateway (with a Pi Zero W) so I found myself thinking that this was perfectably doable with LEGO. Not so cheap, not so discreete and definitely not so autonomous but I already had all I needed so why not?

Another reason was that I learned MQTT a few years ago with the MINDSTORMS EV3. Linux scripting, python, mqtt and BLE... I learned it all with LEGO :D

**Comment by laurensvalk** [comment]
There is now also this [library](https://github.com/pybricks/micropython-bleradio) to further simplify it, and use it with (almost) the same API as on the hubs. This has been an unfinished work for the longest time, so I figured it was finally time to update and publish it :smile:

**Reply by laurensvalk** [reply]
Usage:

```python
# Basic usage of the radio module.

from time import sleep_ms
from bleradio import BLERadio

# A board can broadcast small amounts of data on one channel. Here we broadcast
# on channel 5. This board will listen for other boards on channels 4 and 18.
radio = BLERadio(broadcast_channel=5, observe_channels=[4, 18])

# You can run a variant of this script on another board, and have it broadcast
# on channel 4 or 18, for example. This board will then receive it.

counter = 0

while True:

    # Data observed on channel 4, as broadcast by another board.
    # It gives None if no data is detected.
    observed = radio.observe(4)
    print(observed)

    # Broadcast some data on our channel, which is 5.
    radio.broadcast(["hello, world!", 3.14, counter])
    counter += 1
    sleep_ms(100)
```

**Reply by JorgePe** [reply]
That's great!... but it is still micropython's exclusive.

To the moment I haven't found a way to have micropython with bluetooth  working on RaspberryPi or my Ubuntu laptop. Also never found a pure python library that I could use ('bleak' seems to be client only, found David recommendation to use 'bless' somewhere but it seems also too complex for just advertising data) so this weekend I found a way to do it from command line with just 3 'hcittool' commands.

So importing 'subprocess' and making system calls to 'hcitool' I finally managed to use my Raspberry Pi Zero 2W to send "1" or "0" to the Hubs.

A little more expensive than the Raspberry Pi Pico W but much more stable, my mqtt2pybricks gateway has been running flawless for the last 48 hours.

Unfortunately I read somewhere that 'hcitool' is or is going to be deprecated.

**Reply by laurensvalk** [reply]
> That's great!... but it is still micropython's exclusive.

Yeah, that's been one of the requests (e.g. interface with Arduino). Since you can get a board for $10, that's another way to do it.

We could perhaps do another version with Python. I don't know that all computers can generally advertise, but Raspberry Pi probably can.

For those where this isn't possible, an actual connection may be better, or use a MicroPython board with your computer as above.

**Reply by fkleon** [reply]
@laurensvalk Thanks for publishing that library, looking great!

I found myself in need of a Python implementation too, for debugging purposes and to simulate multiple hubs communicating with each other. So I've been been working on a pure Python, BlueZ-based implementation of the Pybricks BLE broadcast/observe protocol for a while now.

It's been slow going due to time constraints - and a lot of reading about DBus/BlueZ/BLE was required - however I finally managed to clean it up and publish it: https://github.com/fkleon/pybricks-ble

Eventually I'm wanting to offer the same API as on the hubs as well, but for now the observer and broadcaster implementations are separated.

@JorgePe I think this *should* run on a Raspberry Pi with Ubuntu. You're welcome to give it a try. The observer is implemented via `bleak` under the hood. For the broadcaster I did look into `bless` but as you say it's way too complex for simple BLE advertisements, so I implemented this on top of `dbus_fast` (which is what bleak/bless are using under the hood).

**Reply by laurensvalk** [reply]
Looks great @fkleon - thanks for sharing!

**Reply by JorgePe** [reply]
@fkleon Thanks!
I just installed it on a Raspberry Pi Zero 1W with Raspbian OS and used 

`python pb_broadcast --timeout 1 1 1`

to send "1" through channel "1" to my LEGO fingerbot and it worked (the timeout was needed to prevent my robot to react several times).

I just had problems with the 'make format', it kept failing while trying to install 'hidapi'. since I don't understand 'make' and python virtual environments I just forced installation with

`pip install -e '.' --break-system-packages`

(not nice but this is just a test)

I guess I need to check the BlueZ DBus API if I want to mix Pi's and LEGO.

**Comment by VedantKarpe** [comment]
I am trying to broadcast the same from a Raspberry Pi 5. I installed pybricks on raspberry pi and lego hub but the not able to receive the broadcast on my lego spike prime - 

https://docs.pybricks.com/en/latest/hubs/primehub.html#broadcasting-data-to-other-hubs

Any suggestions ?

**Reply by laurensvalk** [reply]
Have you tried [this library](https://pybricks.com/project/micropython-ble-communication/)?

**Reply by VedantKarpe** [reply]
I'll try. This would require micropython right on raspberry pi 5?

**Reply by laurensvalk** [reply]
In principle you could broadcast and observe BLE signals with regular Python or even any other programming language, but we don't have ready-made libraries for that.

But since the protocol is very simple and you can look what the MicroPython library is doing, perhaps you can translate it to your application as needed.

**Reply by JorgePe** [reply]
What a coincidence...
If you intend to use a Raspberry Pi you can use Bleak library (https://github.com/hbldh/bleak) - I just used a RPi Zero 2 with it and a HID library to connect it to my EV3 and pass Pybricks messages to the EV3 as if it is a USB keyboard.
Very easy. So easy that instead of documenting it I decided to try it directly on the EV3... not so easy but yesterday a finally got a EV3 with python 3.11 + bluez 5.72 + latest Bleak receiving messages from my Technic hubs (a decent USB BT 4.0 dongle required of course so the cost is almost the same as using a RPi Zero and much more demanding).

Give me just a couple of days to organize my notes.

**Reply by JorgePe** [reply]
Since documenting is not the best thing I do... this is the script I used with my RPi Zero 2.
It's not smart - just trying to catch messages continuoulsy sent from my Technic hub - every 100 ms it increases a counter and broadcasts it and when I press the button it also broadcasts a string for 700 ms. But you can start from it.

I  am naming all my hubs "Pybricks####". So if 'Pybricks" appears in local_name I know the broadcast is from one of my hubs.
'919' is the BT ID registered by LEGO.

The payload of the message is in the 'manufacturer_data', you can follow Pybricks specs to decode it (first byte is the channel number). Since I'm constantly sending one string with my own custom format I don't need to analyze the 2nd and 3rd bytes and just parse from the 4th byte until the end.

My string format is silly: one byte for identifying the sender,  one byte to describe the message ('C' for counts, 'S' for strings) and the message itself. Since Pybricks already allows sending several types of values it would be better to use it, but before I settled with Bleak I tryed some wild ideas including 'pexpect' to parse data from 'bluetoothctl' command and I really just wanted to receive something at all.

BleakScanner also has an option to assign a callback function. Been using it now but but not in this script.

```python
import asyncio
from bleak import BleakScanner
from zero_hid import Keyboard

async def main():
    while True:
        devices = await BleakScanner.discover(timeout=0.1, return_adv=True)

        if devices is None:
            print('None found')
        else:
            for d in devices:
                dev, advdata = devices[d]
                if 'Pybricks' in str(advdata.local_name):
                    if 919 in advdata.manufacturer_data:
                        payload = list(advdata.manufacturer_data[919])
                        # list(b'\x01\x00\xa31C9')
                        msg_sender = chr(payload[3])
                        msg_type = chr(payload[4])
                        msg_data = bytes(payload[5:])
#                        print(msg_sender, msg_type, end=' | ')
                        if msg_type == 'C':
#                            print( ord(msg_data) )
                            pass
                        elif msg_type == 'S':
                            print( msg_data.decode() )
                            sk.type( msg_data.decode() + '\n')
                            await asyncio.sleep(0.75)
                        else:
                            print('?', msg_data, '?')
                    else:
                        # possible when hub not yet executing
#                        print('Hub but no MD')
                        pass

if __name__ == "__main__":
    sk = Keyboard()
    asyncio.run(main())

```

**Reply by VedantKarpe** [reply]
I'll give it a try and let you know soon.

Update: Was able to use BlueZ to broadcast from my raspberry pi. Was able to detect it on nRF Connect. 
Will try to connect to my lego spike hub tomorrow.

**Reply by JorgePe** [reply]
This is the script I'm now using with a callback function. It works with  python3.11, bleak 0.22.3, dbus-fast 2.39.3 and bluez 5.72 on ev3dev (kernel 4.19.85-ev3dev-3-beta2-ev3) but I didn't test it yet on a Raspberry Pi.

https://github.com/JorgePe/EV3-Pybricks-Messaging/blob/main/test-bleak-v4.py

It's not production ready yet - when switching between channels, I discovered that it tends to cache some messages instead of discarding. You can see this on this video:

[![MINDSTORMS EV3 observing Pybricks messages](http://img.youtube.com/vi/TA4MAE_XB7M/0.jpg)](http://www.youtube.com/watch?v=TA4MAE_XB7M "MINDSTORMS EV3 observing Pybricks messages")

It also works better with less hubs. With more hubs broadcasting at the same time it tends to skip some messages. But for the use I initially intended (a Tic-Tac-Toe robot that just speaks "your turn!" and "you won!") it's already good enough.
